### ISSUE TITLE
Update dnGrep to run on .NET 7 or 8

### ISSUE THREAD

**Issue by doug24** [dnGrep/dnGrep#853]
**Note**: this change is not a commitment at this point, but I am in favor of doing this.  I would like your feedback by voting a thumbs up or thumbs down.  You are welcome to add comments and ask questions about any of this.

Why?
- Search times will be faster - large searches are running 3x faster.
- Move the application to the latest technology and take advantage of improvements to the framework.

Why not?
- Application support for Windows 7 and 8.1 will end.
- A new .NET runtime will be needed, either a separate install or bundled with dnGrep.

Currently dnGrep is built using .NET Framework 4.6.2 and uses the standard Windows installed .NET runtime of 4.6.2, 4.7.x or 4.8.x.   .NET 4.6.2 was released August 2016 and will be supported through 2027, so there is no urgency to move, but all the development has moved to newer versions.  Version 4.8.x of .NET will be supported through the lifetime of Windows 10 and 11.  

Meanwhile, Microsoft has been rewriting a lot of .NET since 2016 and adding all the new features and improvements there.  .NET 7 was released in November 2022, and .NET 8 is planned for November 2023.  They have made significant performance improvements in many areas, and more improvements are expected in .NET8.  The new versions of .NET are only supported on Windows 10 and 11.

I have created a branch of dnGrep running on .NET 7 that I am still testing.  All of the third-party libraries used by dnGrep have been ported to the new .NET except for AlphaFS, which likely will not be ported. AlphaFS is used for long path support, and I have replaced or reimplemented the AlphaFS code.

I created a test script for comparing the performance and search results between the existing dnGrep release and the .NET 7 port.  The script currently runs 10 different searches and creates a report for each.  The existing version dnGrep runs the script in an average of 24 seconds, and the .NET 7 branch takes 8.7 seconds.  Generally, the larger the search, the larger the difference in performance.

The server build process and installation packaging is still work to be done.

Other notes on this proposed change:
- There are no changes to the feature set
- There are no changes to the UI platform (it will still be WPF) or look-and-feel
- There are no changes needed to the language translations
- Once this change is made, the current version on .NET Framework v4 would no longer be supported or updated

In the 3.2.279 release users downloaded:
73% - x64 install kit
19% - portable zip package
8% - x86 install kit

The plan would be to continue support for these three release kits, but the portable package would become x64 only. 

I don't have any visibility for what version of Windows is being used, so I don't know how many Windows 7 users are left.  While I have tried to maintain backward compatibility for Windows 7, I have not had a Windows 7 machine for years now, so there isn't any way for me to test on Windows 7 or reproduce and fix bugs on Windows 7.  

As mentioned above, the new .NET runtimes are no longer included by default in Windows.  There are two options: 
1) Include the full runtime in the install kit or zip file, but this makes the kit much larger ~200MB.
2) Require the user to separately install the runtime from the Microsoft website, and the runtime can be shared by multiple applications.

As an example, the paint.net application is running on .NET 7 and uses the first option of including the full runtime.  

For the .msi install kits, I don't think there is a clear right answer.  I know there are complaints on paint.net about not being able to share the runtime files across applications.

For the portable zip package - I'm guessing that most will prefer a single self-contained set of files so nothing needs to be installed on the host machine.  But note, the zip will be much, much larger.

Please comment if you have any preferences, ideas, or know how other applications have handled this.

**Comment by schittli** [comment]
Hello @doug24 
thanks a lot for all your great work and sharing dnGrep.

In relation to Windows 7 and 8, there are two user groups:

1. Mainly companies that decide to continue using these old operating systems because they have a dependency, e.g. on old hardware.

2. Users who prefer the old operating system, who don't know / don't care how to update it, or who are too lazy to care.

Because there are no more security updates for the old operating systems and hackers are more successful every year and make billions in profits, so it is important that users everywhere realize that it is important and that it is time to update the OS. 

Therefore, the risk increases every day that Windows 7 / 8 users will suffer harm that will lead them to say to themselves: _"Oh!, if only I had updated the OS!"_

So if there are Windows 7 / 8 users who work with dnGrep, it would be a blessing for them if the new dnGrep version motivates them to update the OS.
If they stay with the old OS, then it's good if it hurts them every day because they can't use the new dnGrep features.

Sometimes it is good and important to force people to their happiness in order to protect them from harm.

That's why I would go forward with the .Net framework at all costs and not invest your valuable time in backwards compatibility.

Thanks a lot, kind regards,
Thomas

**Comment by fade2gray** [comment]
What @schittli said. 👍

**Comment by ibay770** [comment]
@doug24 I'm all for the new transition. As mentioned above, I wouldn't worry so much about backward compatibility, and besides, they can always use the previous version which still works well. 
Many thanks for making an awesome program.
i

**Comment by shawnkhall** [comment]
I would think the number of people who use this software on Win7 are rather limited. However, it should be easy enough to do a post-install survey in the next build that just hits a URL with the OS and bit-type of the client. Instead of guessing get the data. Give it a few weeks or a month after the release to tabulate results then you'll know where you really stand.

**Comment by shawnkhall** [comment]
Oh, and for the installer - use the smaller installation package and download (or install from side-by-side) the .NET runtimes during installation only if they're required. There's no need to be "yet another app" that includes all the popular runtimes. Those who need to automate the process can pre-install the runtime or pre-download and save it next to the installation package.

**Comment by doug24** [comment]
Thanks for everyone's comments so far, keep them coming!  

I've been working on the installers, and I agree with the comments here about not having a bundled .NET runtime with the msi kits.  

The tooling for installing the .NET runtime is somewhat lacking at this point: I'm limited to the older WIX Toolset v3.11 by what is installed on the build server, and that version does not have good support for the newer .NET runtimes.  So, I don't think I will have one installer for both the application and the .NET runtime.  Instead, when you start dnGrep the first time, the .NET infrastructure will detect if you have a compatible desktop runtime or not.  If not, it will prompt you to install the correct version.  While it's not seamless, it is very easy to get the right installation.

I'm thinking about having two versions of the 'portable' zip file kits: a large, self-contained version that will run without installing the .NET desktop runtime.  This will be useful for users who do not want anything to installed on the host system. The second portable kit would be like the msi kits with just the dnGrep binaries and will require a separate installation of the .NET desktop runtime.  This is still contingent on the changes needed to the server build, which is my next task.

While testing the install kit on my old laptop, I ran the same search on dnGrep 3.2 and the new dnGrep 4.0.  Here is an example of the improvement I am seeing:

![search-net-framework](https://github.com/dnGrep/dnGrep/assets/17227248/ce64abe6-7224-47a8-9100-54ed530600ea)
![search-net-7](https://github.com/dnGrep/dnGrep/assets/17227248/e5802259-9a81-4789-a16f-05d20c6f6a50)

The .NET 7 version also has much less variation in performance when re-running the same search. The existing 3.2.x version of dnGrep will occasionally run very slowly.

**Comment by DesiQuintans** [comment]
I think that being asked to install runtimes separately is not bad, as long as it's made clear. RStudio requires people to download and install R separately, and they make it very obvious:

![image](https://github.com/dnGrep/dnGrep/assets/3200601/8d489ae2-b0cc-4203-945b-85d9a823ee8a)

**Comment by rickbrew** [comment]
> As an example, the paint.net application is running on .NET 7 and uses the first option of including the full runtime.
> 
> For the .msi install kits, I don't think there is a clear right answer. I know there are complaints on paint.net about not being able to share the runtime files across applications.

There was _one person_ who complained about this, out of millions of PDN users. So I wouldn't worry about it with regard to making this decision.

**Comment by doug24** [comment]
Hi Rick,
Thanks for the comment, I appreciate it!  I know I'm not going to come up with as nice of install and update procedure as you have for PDN.

Just curious if you know - how much did the content of the self-contained .NET files change going from v6 to v7, or may change going from v7 to v8?   Currently dnGrep doesn't use heat.exe to generate the *.wxs files for WIX, and I'll have to do something to manage the file list for a self-contained installer if I go that route.

- Doug

**Comment by rickbrew** [comment]
The .NET runtime file list does change between versions. You don't really need to manually maintain a file list though -- you can just scoop up all the files in e.g. `C:\Program Files\dotnet\shared\Microsoft.NETCore.App\7.0.5` and `C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App\7.0.5`.

If you download the raw nuget packages (which is actually hard to do), e.g. `Microsoft.NETCore.App` and `Microsoft_WindowsDesktop_App`, they contain a `RuntimeList.xml` file which the whole inventory as well as the _type_ of each file. I use this in my packager to do things like dropping the extraneous (and very large) WPF resources DLLs, determine what should be included when running `crossgen2`, etc.

**Comment by reduckted** [comment]
> @doug24  I'm limited to the older WIX Toolset v3.11 by what is installed on the build server

FYI, WiX v4 was released not too long ago and works through NuGet so you don't have to install any special toolset on the build server.

The experience I had upgrading an installer from WiX 3 to WiX 4 was pretty smooth, so I don't think it would be too complicated to update the installer for dnGrep.

**Comment by doug24** [comment]
@reduckted - Thanks! I didn't know it was a NuGet package now, that explains why it's not on AppVeyor's tool list.

**Comment by doug24** [comment]
Status - 

My current plan is to have both the msi and zip builds self-contained, so they will include the .NET runtime with the application.  I went this way because I think the portable (zip) builds need to be self-contained, so nothing needs to be installed. Making the msi also self-contained means there doesn't need to separate builds for msi and zip.  The app will need to be updated when Microsoft issues security patches for .NET.  This also means the user doesn't need to install the .NET runtime separately.  

For now I'm staying with Wix 3.11.  I tried the upgrade to Wix 4, but ran into a blocking issue that is supposed to be fixed in an upcoming update.

I hope to have a release candidate out for you to test today or tomorrow.

**Comment by doug24** [comment]
The release candidate 1 is published here: https://github.com/dnGrep/dnGrep/releases/tag/v4.0.8.0

Please test and report any issues.

**Comment by kintrupf** [comment]
Adding 60 MB for a .NET runtime that almost always already present on the user's machine is quite a big step.
I can follow the reasoning behind your decision, but it's still a lot of overhead.

I Installed it on Windows 10. Something with the result list is not working. 
It does not show any results, only a white rectangle with a very big gray border:

![image](https://github.com/dnGrep/dnGrep/assets/8496344/d40d0767-6ede-4c9c-890b-f1c2b34e8a78)

**Comment by doug24** [comment]
That's not good.  I don't know what the problem could be there.  I do have it installed on both Windows 10 and 11, and it's working for me.  What version of Windows 10 do you have?

**Comment by yyjdelete** [comment]
Consider use `StringComparison.Ordinal/OrdinalIgnoreCase` instead of `StringComparison.InvariantCulture/InvariantCultureIgnoreCase` for text search to get better performance?
https://github.com/dnGrep/dnGrep/blob/88da3e4721bce392ab99b217368e6c1a3fe357da/dnGREP.Engines/GrepEngineBase.cs#L448

text, case sensitive=on, parallel=off
11.684s(InvariantCulture)/6.523s(Ordinal)
text, case sensitive=off, parallel=off
12.159s(InvariantCultureIgnoreCase)/6.956s(OrdinalIgnoreCase)

parallel=on(almost the same, but Ordinal ones with less cpu fan speed and cpu usage)
text, case sensitive=on, parallel=on
2.955s(InvariantCulture)/2.541s(Ordinal)
text, case sensitive=off, parallel=on
3.138s(InvariantCultureIgnoreCase)/2.555s(OrdinalIgnoreCase)

**Comment by doug24** [comment]
@yyjdelete I agree it should not be InvariantCulture, that's not right.  And Ordinal - binary matching - will be faster.  But I wonder if CurrentCulture should be used (which would be slow)? Using CurrentCulture takes into account character sequences that are equivalent in some languages, such as strasse and straße in German, or different ways of using Unicode combining characters. I'm surprised that using culture comparisons has not been an issue before.  Maybe we should make Ordinal the default and add a new option for CurrentCulture.

**Comment by kintrupf** [comment]
> @kintrupf - there have not been any more reports of this problem, so I'm wondering if it's something in your environment. If you have not done so already, can you try:
> 
>     * uninstall, make sure the installation directory is deleted or empty, and reinstall.
> 
>     * close dnGrep, rename or move your `dnGREP.Settings.dat` file (in AppData\Roaming\dnGREP), and restart?

I checked it on another Windows 10 system and renamed the user folder ...\AppData\Roaming\dnGREP before I installed it. First impression is that the new version seems to work without any glitches.

**Comment by kintrupf** [comment]
> @yyjdelete I agree it should not be InvariantCulture, that's not right. And Ordinal - binary matching - will be faster. But I wonder if CurrentCulture should be used (which would be slower than Ordinal)? Using CurrentCulture takes into account character sequences that are equivalent in some languages, such as strasse and straße in German, or different ways of using Unicode combining characters. I'm surprised that using culture comparisons has not been an issue before. Maybe we should make Ordinal the default and add a new option for CurrentCulture.

Is it possible to make this dependent on the search string? Maybe it's okay to used the faster "Ordinal" (binary) search as default if only standard characters (ASCII 32-127 or something like that) are used in the search string and "CurrentCulture" if something else is present?

**Comment by yyjdelete** [comment]
> > @yyjdelete I agree it should not be InvariantCulture, that's not right. And Ordinal - binary matching - will be faster. But I wonder if CurrentCulture should be used (which would be slower than Ordinal)? Using CurrentCulture takes into account character sequences that are equivalent in some languages, such as strasse and straße in German, or different ways of using Unicode combining characters. I'm surprised that using culture comparisons has not been an issue before. Maybe we should make Ordinal the default and add a new option for CurrentCulture.
> 
> Is it possible to make this dependent on the search string? Maybe it's okay to used the faster "Ordinal" (binary) search as default if only standard characters (ASCII 32-127 or something like that) are used in the search string and "CurrentCulture" if something else is present?

It's really usage and user language dependent, if culture search is required, maybe it's better to give an option to user.

And standard characters is also user langage dependent, for Chinese(zh-CN, with both UTF8 or GBK encoding), standard characters mean printable latin:(\t, 0x20~0x7E), chinese chars:(unicode)0x3400-0x4DBF, 0x4E00-0x9FBF,0xF900-0xFA2D,(ext unicode)0x20000-0x2A6DF
>And what's more, for example, the culture info say 一,㊀,㆒are same, but most people in China don't think so, so do ordinal compare is enough in China.

And when search function or var name from code,  ordinal is enough in most case.

**Comment by doug24** [comment]
I am changing the default text search to use Ordinal.  I'm still considering the option for CurrentCulture - I would rather not make another option is it's not needed.  No one has ever raised an issue about culture specific matches, but I think not many users would be aware of the possibility.

I'm targeting June 14 or 15 to make this, with updates, the full release (not another pre-release).

**Comment by mi-hofmann** [comment]
Hi,

I am having the same issue:
> I Installed it on Windows 10. Something with the result list is not working. It does not show any results, only a white rectangle with a very big gray border:
> 
> ![image](https://user-images.githubusercontent.com/8496344/242981432-d40d0767-6ede-4c9c-890b-f1c2b34e8a78.png)

But without the white rectangle.

What did not help:
* uninstall dnGrep and delete `%PROGRAMFILES%\dnGrep`

What *did* help:
* delete `%APPDATA%\dnGrep\dnGREP.Settings.dat`

I don't know which of the settings files' content is causing the issue, but I trimmed down my old (bad) settings file a bit and can reproduce the problem with that file:
[bad_dnGREP.Settings.zip](https://github.com/dnGrep/dnGrep/files/12417482/bad_dnGREP.Settings.zip)

By the way, version 4 is incredibly much faster, wow! Thank you very much!

**Comment by doug24** [comment]
You're welcome!  And thanks for the detective work!  

In your language/culture conventions do you use comma for the decimal separator?  If so, I have fixed a bug that was writing numeric values in the local culture and reading them using a .net 'invariant' culture where a dot is the decimal separator.  That would explain how the value got so large.

**Comment by mi-hofmann** [comment]
That makes sense. I am currently melting under a hot german sun, where `,` and `.` are used the other way around. So yes, I guess.
